File: ecos-records/src/main/java/ru/citeck/ecos/records3/record/op/atts/service/schema/SchemaAtt.java
package ru.citeck.ecos.records3.record.op.atts.service.schema;

import lombok.Data;
import lombok.ToString;
import ru.citeck.ecos.commons.utils.MandatoryParam;
import ru.citeck.ecos.records3.record.op.atts.service.proc.AttProcDef;
import ru.citeck.ecos.records3.record.op.atts.service.schema.exception.AttSchemaException;

import java.util.*;
import java.util.stream.Collectors;

@Data
public class SchemaAtt {

    private final String alias;
    private final String name;
    private final boolean multiple;
    private final List<SchemaAtt> inner;
    private final List<AttProcDef> processors;

    public static Builder create() {
        return new Builder();
    }

    private SchemaAtt(String alias,
                      String name,
                      boolean multiple,
                      List<SchemaAtt> inner,
                      List<AttProcDef> processors) {

        this.alias = alias;
        this.name = name;
        this.multiple = multiple;
        this.inner = inner;
        this.processors = processors;
    }

    public Builder copy() {
        return new Builder(this);
    }

    public String getScalarName() {
        if (isScalar()) {
            return getName();
        } else {
            if (inner.size() != 1) {
                return null;
            }
            return inner.get(0).getScalarName();
        }
    }

    public boolean isScalar() {
        return name.length() > 0 && name.charAt(0) == '?';
    }

    public String getAliasForValue() {
        if (alias.isEmpty()) {
            return name;
        }
        return alias;
    }

    @Override
    public String toString() {

        String res = "{";
        if (!alias.isEmpty()) {
            res += "\"alias\":\"" + alias + "\",";
        }
        res += "\"name\":\"" + name + '\"'
            + ", \"multiple\":" + multiple;

        if (!inner.isEmpty()) {
            res += ", \"inner\":[" + inner.stream()
                .map(SchemaAtt::toString)
                .collect(Collectors.joining(", ")) + "]";
        }
        if (!processors.isEmpty()) {
            res += ", \"processors\":[" + processors.stream()
                .map(AttProcDef::toString)
                .collect(Collectors.joining(", ")) + "]";
        }
        return res + '}';
    }

    @ToString
    public static class Builder {

        private String alias = "";
        private String name;
        private boolean multiple;
        private List<SchemaAtt> inner = Collections.emptyList();
        private List<AttProcDef> processors = Collections.emptyList();

        public Builder() {
        }

        public Builder(SchemaAtt base) {
            this.alias = base.getAlias();
            this.name = base.getName();
            this.multiple = base.isMultiple();
            this.inner = base.getInner();
            this.processors = base.getProcessors();
        }

        public Builder setAlias(String alias) {
            this.alias = alias;
            return this;
        }

        public Builder setName(String name) {
            this.name = name;
            return this;
        }

        public Builder setMultiple(boolean multiple) {
            this.multiple = multiple;
            return this;
        }

        public boolean isMultiple() {
            return this.multiple;
        }

        public List<SchemaAtt> getInner() {
            return inner;
        }

        public Builder setInner(List<SchemaAtt> inner) {
            this.inner = inner;
            return this;
        }

        public Builder setInner(SchemaAtt inner) {
            return setInner(Collections.singletonList(inner));
        }

        public Builder setInner(SchemaAtt.Builder inner) {
            return setInner(inner.build());
        }

        public Builder setProcessors(List<AttProcDef> processors) {
            if (processors == null) {
                this.processors = Collections.emptyList();
            } else {
                this.processors = new ArrayList<>(processors);
            }
            return this;
        }

        public SchemaAtt build() {

            MandatoryParam.check("name", name);

            SchemaAtt att = new SchemaAtt(
                alias,
                name,
                multiple,
                Collections.unmodifiableList(inner),
                Collections.unmodifiableList(processors)
            );

            if (att.isScalar() && !inner.isEmpty()) {
                throw new AttSchemaException("Attribute can't be a scalar and has inner attributes. " + this);
            }
            if (!att.isScalar() && inner.isEmpty()) {
                throw new AttSchemaException("Attribute can't be not a scalar and has empty inner attributes. " + this);
            }
            if (att.isScalar() && att.isMultiple()) {
                throw new AttSchemaException("Scalar can't hold multiple values. " + this);
            }
            if (att.isScalar() && !ScalarType.getBySchema(att.getName()).isPresent()) {
                throw new AttSchemaException("Unknown scalar: '" + att.getName() + "'");
            }

            return att;
        }
    }
}

File: ecos-records/src/main/java/ru/citeck/ecos/records3/record/op/atts/service/schema/SchemaAtt.kt
package ru.citeck.ecos.records3.record.op.atts.service.schema

import ecos.com.fasterxml.jackson210.annotation.JsonSetter
import ecos.com.fasterxml.jackson210.databind.annotation.JsonDeserialize
import ru.citeck.ecos.commons.utils.MandatoryParam
import ru.citeck.ecos.records3.record.op.atts.service.proc.AttProcDef
import ru.citeck.ecos.records3.record.op.atts.service.schema.exception.AttSchemaException
import ru.citeck.ecos.records3.record.op.atts.service.schema.write.AttSchemaGqlWriter
import com.fasterxml.jackson.annotation.JsonSetter as JackJsonSetter
import com.fasterxml.jackson.databind.annotation.JsonDeserialize as JackJsonDeserialize

@JsonDeserialize(builder = SchemaAtt.Builder::class)
@JackJsonDeserialize(builder = SchemaAtt.Builder::class)
data class SchemaAtt(
    val alias: String,
    val name: String,
    val multiple: Boolean,
    val inner: List<SchemaAtt>,
    val processors: List<AttProcDef>
) {

    companion object {

        @JvmStatic
        fun create(): Builder {
            return Builder()
        }

        @JvmStatic
        fun create(builder: Builder.() -> Unit): SchemaAtt {
            val builderObj = Builder()
            builder.invoke(builderObj)
            return builderObj.build()
        }
    }

    fun getScalarName(): String? {
        return if (isScalar()) {
            name
        } else {
            if (inner.size != 1) {
                null
            } else inner[0].getScalarName()
        }
    }

    fun isScalar(): Boolean {
        return name.isNotEmpty() && name[0] == '?'
    }

    fun getAliasForValue(): String {
        return if (alias.isEmpty()) {
            name
        } else {
            alias
        }
    }

    fun copy(): Builder {
        return Builder(this)
    }

    fun copy(builder: Builder.() -> Unit): SchemaAtt {
        val builderObj = Builder(this)
        builder.invoke(builderObj)
        return builderObj.build()
    }

    override fun toString(): String {
        return AttSchemaGqlWriter.INSTANCE.write(this)
    }

    class Builder() {

        var alias: String = ""
        lateinit var name: String
        var multiple: Boolean = false
        var inner: List<SchemaAtt> = emptyList()
        var processors: List<AttProcDef> = emptyList()

        constructor(base: SchemaAtt) : this() {
            this.alias = base.alias
            this.name = base.name
            this.multiple = base.multiple
            this.inner = base.inner
            this.processors = base.processors
        }

        fun withAlias(alias: String): Builder {
            this.alias = alias
            return this
        }

        fun withName(name: String): Builder {
            this.name = name
            return this
        }

        fun withMultiple(multiple: Boolean): Builder {
            this.multiple = multiple
            return this
        }

        fun withInner(inner: SchemaAtt.Builder): Builder {
            return withInner(inner.build())
        }

        fun withInner(inner: SchemaAtt): Builder {
            return withInner(listOf(inner))
        }

        @JsonSetter
        @JackJsonSetter
        fun withInner(inner: List<SchemaAtt>): Builder {
            this.inner = inner
            return this
        }

        fun withProcessors(processors: List<AttProcDef>): Builder {
            this.processors = processors
            return this
        }

        fun build(): SchemaAtt {

            MandatoryParam.check("name", name)
            val att = SchemaAtt(alias, name, multiple, inner, processors)

            if (att.isScalar() && !inner.isEmpty()) {
                throw AttSchemaException("Attribute can't be a scalar and has inner attributes. $this")
            }
            if (!att.isScalar() && inner.isEmpty()) {
                throw AttSchemaException("Attribute can't be not a scalar and has empty inner attributes. $this")
            }
            if (att.isScalar() && att.multiple) {
                throw AttSchemaException("Scalar can't hold multiple values. $this")
            }
            if (att.isScalar() && !ScalarType.getBySchema(att.name).isPresent) {
                throw AttSchemaException("Unknown scalar: '" + att.name + "'")
            }

            return att
        }
    }
}
